clientpool/channel.go

```diff
@@ -19,25 +22,55 @@ var _ Pool = (*channelPool)(nil)
 // NewChannelPool creates a new client pool implemented via channel.
 //
 // Note that this function could return both non-nil Pool and error,
-// when we failed to create all asked initialClients.
-// In such case the returned Pool would have the clients we already established.
-func NewChannelPool(initialClients, maxClients int, opener ClientOpener) (Pool, error) {
-	if initialClients > maxClients {
+// when we failed to create all asked bestEffortInitialClients.
+// In such case the returned Pool would have at least requiredInitialClients
+// clients we already established.
+func NewChannelPool(ctx context.Context, requiredInitialClients, bestEffortInitialClients, maxClients int, opener ClientOpener) (Pool, error) {
+	if !(requiredInitialClients <= bestEffortInitialClients && bestEffortInitialClients <= maxClients) {
 		return nil, &ConfigError{
-			InitialClients: initialClients,
-			MaxClients:     maxClients,
+			BestEffortInitialClients: bestEffortInitialClients,
+			RequiredInitialClients:   requiredInitialClients,
+			MaxClients:               maxClients,
 		}
 	}
 
-	var finalErr error
+	var lastAttemptErr error
 	pool := make(chan Client, maxClients)
-	for i := 0; i < initialClients; i++ {
+
+	for i := 0; i < requiredInitialClients; {
+		if ctxErr := ctx.Err(); ctxErr != nil {
+			if lastAttemptErr == nil {
+				// In case the user sets a deadline so short that we don't have
+				// time to open all the client serially despite all of them working.
+				// In that case lastAttempErr would be nil so we need to indicate to
+				// the user that their timeout being too short is the issue.
+				lastAttemptErr = ctxErr
+			}
+			return &channelPool{
+				pool:           pool,
+				opener:         opener,
+				initialClients: len(pool),
+				maxClients:     maxClients,
+			}, lastAttemptErr
+		}
+		c, err := opener()
+		if err == nil {
+			pool <- c
+			i++
+		} else {
+			lastAttemptErr = err
+			log.Warnf("clientpool: error creating required client (will retry): %w", err)
```

Comment:
I don't know that I would add a `log.Warn` here, I would worry about this causing a lot of log spam if there's an issue?

Reply:
@fishy really wanted a way to make the error clear to users.

Reply:
added at 2/s